File: Treachery-iOS/Treachery-iOS/Models/Player.swift
//
//  Player.swift
//  Treachery-iOS
//
//  Created by Luke Solomon on 3/12/26.
//

import Foundation

struct Player: Codable, Identifiable, Equatable {
    var id: String
    let orderId: Int
    let userId: String
    var displayName: String
    var role: Role?
    var identityCardId: String?
    var lifeTotal: Int
    var isEliminated: Bool
    var isUnveiled: Bool
    let joinedAt: Date
    var playerColor: String?
    var commanderName: String?
    var originalIdentityCardId: String?
    var isFaceDown: Bool
    var isReady: Bool

    enum CodingKeys: String, CodingKey {
        case id
        case orderId = "order_id"
        case userId = "user_id"
        case displayName = "display_name"
        case role
        case identityCardId = "identity_card_id"
        case lifeTotal = "life_total"
        case isEliminated = "is_eliminated"
        case isUnveiled = "is_unveiled"
        case joinedAt = "joined_at"
        case playerColor = "player_color"
        case commanderName = "commander_name"
        case originalIdentityCardId = "original_identity_card_id"
        case isFaceDown = "is_face_down"
        case isReady = "is_ready"
    }

    // Custom decoder: uses decodeIfPresent for 'id' to handle documents
    // created by other clients (e.g. Android) that store the player ID
    // only as the Firestore document ID, not as a field in the data.
    // The actual document ID is injected after decoding in FirestoreManager.
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(String.self, forKey: .id) ?? ""
        orderId = try container.decode(Int.self, forKey: .orderId)
        userId = try container.decode(String.self, forKey: .userId)
        displayName = try container.decode(String.self, forKey: .displayName)
        role = try container.decodeIfPresent(Role.self, forKey: .role)
        identityCardId = try container.decodeIfPresent(String.self, forKey: .identityCardId)
        lifeTotal = try container.decode(Int.self, forKey: .lifeTotal)
        isEliminated = try container.decode(Bool.self, forKey: .isEliminated)
        isUnveiled = try container.decode(Bool.self, forKey: .isUnveiled)
        joinedAt = try container.decode(Date.self, forKey: .joinedAt)
        playerColor = try container.decodeIfPresent(String.self, forKey: .playerColor)
        commanderName = try container.decodeIfPresent(String.self, forKey: .commanderName)
        originalIdentityCardId = try container.decodeIfPresent(String.self, forKey: .originalIdentityCardId)
        isFaceDown = try container.decodeIfPresent(Bool.self, forKey: .isFaceDown) ?? false
        isReady = try container.decodeIfPresent(Bool.self, forKey: .isReady) ?? false
    }

    init(
        id: String,
        orderId: Int,
        userId: String,
        displayName: String,
        role: Role?,
        identityCardId: String?,
        lifeTotal: Int,
        isEliminated: Bool,
        isUnveiled: Bool,
        joinedAt: Date,
        playerColor: String? = nil,
        commanderName: String? = nil,
        originalIdentityCardId: String? = nil,
        isFaceDown: Bool = false,
        isReady: Bool = false
    ) {
        self.id = id
        self.orderId = orderId
        self.userId = userId
        self.displayName = displayName
        self.role = role
        self.identityCardId = identityCardId
        self.lifeTotal = lifeTotal
        self.isEliminated = isEliminated
        self.isUnveiled = isUnveiled
        self.joinedAt = joinedAt
        self.playerColor = playerColor
        self.commanderName = commanderName
        self.originalIdentityCardId = originalIdentityCardId
        self.isFaceDown = isFaceDown
        self.isReady = isReady
    }
}

File: TreacheryAndroid/app/src/main/java/com/solomon/treachery/model/Player.kt
package com.solomon.treachery.model

import com.google.firebase.Timestamp

data class Player(
    val id: String = "",
    val orderId: Int = 0,
    val userId: String = "",
    val displayName: String = "",
    val role: Role? = null,
    val identityCardId: String? = null,
    val lifeTotal: Int = 40,
    val isEliminated: Boolean = false,
    val isUnveiled: Boolean = false,
    val joinedAt: Timestamp = Timestamp.now(),
    val playerColor: String? = null,
    val commanderName: String? = null,
    val originalIdentityCardId: String? = null,
    val isFaceDown: Boolean = false
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "id" to id,
        "order_id" to orderId,
        "user_id" to userId,
        "display_name" to displayName,
        "role" to role?.value,
        "identity_card_id" to identityCardId,
        "life_total" to lifeTotal,
        "is_eliminated" to isEliminated,
        "is_unveiled" to isUnveiled,
        "joined_at" to joinedAt,
        "player_color" to playerColor,
        "commander_name" to commanderName,
        "original_identity_card_id" to originalIdentityCardId,
        "is_face_down" to isFaceDown
    )

    companion object {
        fun fromMap(id: String, data: Map<String, Any?>): Player = Player(
            id = id,
            orderId = (data["order_id"] as? Number)?.toInt() ?: 0,
            userId = data["user_id"] as? String ?: "",
            displayName = data["display_name"] as? String ?: "",
            role = (data["role"] as? String)?.let { Role.fromValue(it) },
            identityCardId = data["identity_card_id"] as? String,
            lifeTotal = (data["life_total"] as? Number)?.toInt() ?: 40,
            isEliminated = data["is_eliminated"] as? Boolean ?: false,
            isUnveiled = data["is_unveiled"] as? Boolean ?: false,
            joinedAt = data["joined_at"] as? Timestamp ?: Timestamp.now(),
            playerColor = data["player_color"] as? String,
            commanderName = data["commander_name"] as? String,
            originalIdentityCardId = data["original_identity_card_id"] as? String,
            isFaceDown = data["is_face_down"] as? Boolean ?: false
        )
    }
}
